### DISCUSSION TITLE
Trait bound is not satisfied when attempting to use with_state()

### DISCUSSION THREAD

**Discussion by AnthonysDevAccount** [tokio-rs/axum#2713] [Q&A] [answered]
### Summary

Hi.  I get the following errors when running cargo check:

error[E0277]: the trait bound `for<'a> Router<AppState>: tower_service::Service<IncomingStream<'a>>` is not satisfied
  --> src\main.rs:13:27
   |
13 |     axum::serve(listener, app).await.unwrap();
   |     -----------           ^^^ the trait `for<'a> tower_service::Service<IncomingStream<'a>>` is not implemented for `Router<AppState>`
   |     |
   |     required by a bound introduced by this call
   |
   = help: the following other types implement trait `tower_service::Service<Request>`:
             <Router as tower_service::Service<axum::http::Request<B>>>
             <Router as tower_service::Service<IncomingStream<'_>>>
note: required by a bound in `serve`
  --> C:\Users\admin\.cargo\registry\src\index.crates.io-6f17d22bba15001f\axum-0.7.5\src\serve.rs:97:8
   |
95 | pub fn serve<M, S>(tcp_listener: TcpListener, make_service: M) -> Serve<M, S>
   |        ----- required by a bound in this function
96 | where
97 |     M: for<'a> Service<IncomingStream<'a>, Error = Infallible, Response = S>,
   |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `serve`

error[E0277]: the trait bound `Serve<Router<AppState>, _>: IntoFuture` is not satisfied
  --> src\main.rs:13:32
   |
13 |     axum::serve(listener, app).await.unwrap();
   |     ---------------------------^^^^^
   |     |                         ||
   |     |                         |the trait `IntoFuture` is not implemented for `Serve<Router<AppState>, _>`
   |     |                         help: remove the `.await`
   |     this call returns `Serve<Router<AppState>, _>`
   |
   = help: the trait `IntoFuture` is implemented for `Serve<M, S>`

For more information about this error, try `rustc --explain E0277`.
error: could not compile `web` (bin "web") due to 2 previous errors

I'm new to Axum so I apologise if the solution is obvious.  Any help would be appreciated.

Following is the source code:

use axum::routing::get;

#[tokio::main]
async fn main() {
    let shared_state: AppState  = AppState { num: 42 };
    let app = setup_routing(shared_state); 
    let listener = tokio::net::TcpListener::bind("localhost:5000").await.unwrap();
    axum::serve(listener, app).await.unwrap();
}

#[allow(dead_code)]
fn setup_routing(shared_state: AppState) -> axum::Router<AppState>  {
    axum::Router::new()
        .fallback(fallback)
        .nest_service("/static", tower_http::services::ServeDir::new("static"))
        .route("/", get(get_root))
        .route("/show-state", get(show_state))
        .with_state::<AppState>(shared_state) // only adds state to all prior routes 
}

async fn show_state(axum::extract::State(shared_state): axum::extract::State<AppState>) -> String {
    format!("shared state = {shared_state:?}")
}

async fn fallback(uri: axum::http::Uri) -> impl axum::response::IntoResponse {
    (axum::http::StatusCode::NOT_FOUND, format!("No route {uri}"))
}

async fn get_root() -> axum::response::Html<&'static str> {
    include_str!("../static/root.html").into()
}

#[derive(Clone, Debug, Copy)]
#[allow(dead_code)]
pub struct AppState {
    num: i32,
}

### axum version

0.7.5

**Comment by wangziling** [comment, marked as answer]
Hi you need to remove the `turbo fish` (double colons) next to the `.with_state`.
And then you need to point the `axum::extract::State`'s generic type thru `axum::extract::State<AppState>`.

![image](https://github.com/tokio-rs/axum/assets/24818903/e073250c-3080-4911-8980-4cb07577064b)

The final code is below:

```rs
use axum::routing::get;

#[tokio::main]
async fn main() {
    let shared_state: AppState = AppState { num: 42 };
    let app = setup_routing(shared_state);
    let listener = tokio::net::TcpListener::bind("localhost:5000")
        .await
        .unwrap();
    axum::serve(listener, app).await.unwrap();
}

#[allow(dead_code)]
fn setup_routing(shared_state: AppState) -> axum::Router {
    axum::Router::new()
        .fallback(fallback)
        .nest_service("/static", tower_http::services::ServeDir::new("static"))
        .route("/", get(get_root))
        .route("/show-state", get(show_state))
        .with_state(shared_state) // only adds state to all prior routes
}

async fn show_state(axum::extract::State(shared_state): axum::extract::State<AppState>) -> String {
    format!("shared state = {shared_state:?}")
}

async fn fallback(uri: axum::http::Uri) -> impl axum::response::IntoResponse {
    (axum::http::StatusCode::NOT_FOUND, format!("No route {uri}"))
}

async fn get_root() -> axum::response::Html<&'static str> {
    include_str!("../static/root.html").into()
}

#[derive(Clone, Debug, Copy)]
#[allow(dead_code)]
pub struct AppState {
    num: i32,
}

```

**Comment by AnthonysDevAccount** [comment]
Hi.  That worked!
Thanks for your help.
